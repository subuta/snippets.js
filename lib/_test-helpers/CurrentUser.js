import { build, format, snippets as s } from 'bld.js'

export default (sub = 'google-oauth2|dummy') => {
  return build`
    export const createPayload = (sub) => ({
      sub,
      iss: 'https://xxx.com/',
      aud: ['https://xxx.com/api', 'https://xxx.auth0.com/userinfo'],
      scope: 'openid profile email'
    })
    
    export const currentUser = createPayload(${s.stringify(sub)})
  `
}
