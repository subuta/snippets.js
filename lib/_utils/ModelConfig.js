import check from 'lib/_utils/apiCheck'

const ModelConfigShape = check.shape({
  schema: check.shape({
    tableName: check.string,
    required: check.arrayOf(check.string).optional,
    properties: check.objectOf(check.shape({
      type: check.string,
    })),

    relations: check.objectOf(check.shape({
      hasMany: check.string.optional,
      join: check.shape({
        from: check.string,
        to: check.string
      })
    })).optional
  }),
  // mark as junction-table.
  isJunction: check.bool.optional
})

// eg:
// config = {
//   book: {
//     schema: {}
//   }
// }
export const validateModels = function (config) {
  return check.throw(check.objectOf(ModelConfigShape), config)
}

// eg:
// config = {
//   schema: {}
// }
export const validateModel = function (config) {
  return check.throw(ModelConfigShape, config)
}
