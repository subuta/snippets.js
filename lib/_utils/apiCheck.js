import apiCheck from 'api-check'

export default apiCheck({
    output: {
        docsBaseUrl: 'https://github.com/subuta/snippets.js#'
    },
    verbose: false
});
