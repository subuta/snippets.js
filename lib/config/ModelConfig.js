import check from 'lib/_utils/apiCheck'

const ModelConfigShape = check.shape({
  schema: check.object
})

// eg:
// config = {
//   book: {
//     schema: {}
//   }
// }
export const validateModels = function (config) {
  return check.throw(check.objectOf(ModelConfigShape), config)
}

// eg:
// config = {
//   schema: {}
// }
export const validateModel = function (config) {
  return check.throw(ModelConfigShape, config)
}
