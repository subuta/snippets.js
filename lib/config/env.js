import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'

export default (imports = []) => {
  const vars = _.zipObject(imports, imports)

  return build`
    // load .env first.
    require('dotenv').config()
    
    const ${s.raw(vars)} = process.env
    
    export ${s.raw(vars)}
    
    export default ${s.raw(vars)}
  `
}
