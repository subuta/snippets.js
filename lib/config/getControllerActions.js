import _ from 'lodash'

export default (config = {}) => {
  const {
    only, // the routes will generated.
    except  // the routes will ignored.
  } = config

  const routes = [
    'index',
    'show',
    'create',
    'update',
    'destroy'
  ]

  if (only) {
    return _.intersection(routes, only)
  } else if (except) {
    return _.difference(routes, except)
  }

  return routes
}
