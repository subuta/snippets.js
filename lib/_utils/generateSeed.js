import {validateModel} from 'lib/_utils/ModelConfig'
import _ from 'lodash'
import faker from 'faker'
import fakerEn from 'faker/locale/en_US'
import { snippets as s } from 'bld.js'

import hash from 'string-hash'

export default (modelConfig, locale = 'en') => {
  validateModel(modelConfig)

  // use hash of json as a randomness
  faker.seed(hash(JSON.stringify(modelConfig)));

  // set locale.
  faker.locale = locale

  let {tableName, properties} = modelConfig.schema
  tableName = s.tableName(tableName)

  return _.transform(properties, (result, prop, key) => {
    result[key] = (function generator (prop, key) {
      const {type} = prop
      const _key = _.snakeCase(key)

      // generate data for well-known column.
      if (_key === 'zip_code') {
        return faker.address.zipCode()
      }

      if (type === 'string' && _.includes(_key, 'address')) {
        return faker.address.streetAddress()
      }

      if (_.includes(_key, 'city')) {
        return faker.address.city()
      }

      if (_.endsWith(_key, 'price')) {
        return faker.commerce.price()
      }

      if (_.includes(_key, 'state')) {
        return faker.address.state()
      }

      if (_.includes(_key, 'country_code')) {
        return faker.address.countryCode()
      }

      if (_.includes(_key, 'country')) {
        return faker.address.country()
      }

      if (_key === 'last_name') {
        return faker.name.lastName()
      }

      if (_.includes(_key, 'color')) {
        return faker.commerce.color()
      }

      if (_key === 'first_name') {
        return faker.name.firstName()
      }

      if (_.includes(_key, 'phone') || _.startsWith(_key, 'tel')) {
        return faker.phone.phoneNumber()
      }

      if (_.includes(_key, 'mail')) {
        return fakerEn.internet.email()
      }

      if (_.includes(_key, 'tag') || _.includes(s.singular(_key), 'category')) {
        return [faker.random.word()]
      }

      if (_.startsWith(_key, 'last') && _.endsWith(_key, 'at')) {
        return faker.date.past()
      }

      if (key.match(/starts?_at/)) {
        return faker.date.past()
      }

      if (key.match(/ends?_at/)) {
        return faker.date.future()
      }

      if (_.endsWith(_key, 'at') || _.endsWith(_key, 'date')) {
        return faker.date.future()
      }

      if (_key === 'uuid') {
        return faker.random.uuid()
      }

      if (_.endsWith(_key, 'url')) {
        return faker.internet.url()
      }

      if (_key === 'id' || _.endsWith(_key, 'id')) {
        return faker.random.number()
      }

      if (_.endsWith(_key, 'count') || _.endsWith(_key, 'counts')) {
        return faker.random.number()
      }

      if (_key === 'locale') {
        return faker.random.locale()
      }

      if (_.includes(_key, 'avatar')) {
        return faker.internet.avatar()
      }

      if (_.includes(_key, 'image')) {
        return faker.image.image()
      }

      if (_.includes(tableName, 'attachment') && (_key === 'url' || _key === 'image')) {
        return faker.image.cats()
      }

      if (_.includes(tableName, 'companies') && (_key === 'name' || _key === 'label')) {
        return faker.company.companyName()
      }

      if (_.includes(tableName, 'user') && (_key === 'name')) {
        if (locale === 'ja') return `${faker.name.lastName()} ${faker.name.firstName()}`

        return faker.name.findName()
      }

      if ((_.includes(tableName, 'product') || _.includes(tableName, 'item')) && _key === 'name') {
        return faker.commerce.productName()
      }

      if (_key === 'user_name' || _key === 'username') {
        return fakerEn.internet.userName()
      }

      // comment
      if (_key === 'comment' || (_.includes(tableName, 'comment') && _key === 'text')) {
        return fakerEn.lorem.paragraph()
      }

      if (_key === 'content' || _key === 'body') {
        return fakerEn.lorem.paragraph()
      }

      // string specific conditions
      if (type === 'string' && prop.format) {
        switch (prop.format) {
          case 'date':
          case 'date-time':
            return faker.date.future().toISOString()
          case 'uri':
            return faker.internet.url()
          case 'email':
            return fakerEn.internet.email()
          case 'ipv4':
            return faker.internet.ip()
          case 'ipv6':
            return faker.internet.ipv6()
        }
      }

      // then try to generate by column-type.
      switch (type) {
        case 'array':
          return [generator(prop.items, key)]
        case 'object':
          return _.transform(prop.properties || {}, (result, p, k) => {
            result[k] = generator(p, k)
          }, {})
        case 'number':
        case 'integer':
          return faker.random.number()
        case 'boolean':
          return faker.random.boolean()
        default:
          return faker.random.words()
      }
    })(prop, key)
  }, {})
}
