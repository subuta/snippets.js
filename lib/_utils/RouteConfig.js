import check from 'lib/_utils/apiCheck'

import {CONTROLLER_ACTIONS} from 'lib/_utils/getControllerActions'

const RouteConfigShape = check.shape({
  except: check.arrayOf(check.oneOf(CONTROLLER_ACTIONS)).optional,
  only: check.arrayOf(check.oneOf(CONTROLLER_ACTIONS)).optional,
  skipAuth: check.oneOfType([
    check.bool,
    check.arrayOf(check.oneOf(CONTROLLER_ACTIONS))
  ]).optional,
  imports: check.array.optional,
  eager: check.string.optional,
  prefix: check.string.optional,
})

// eg:
// config = {
//   book: {
//     except: ['index']
//   }
// }
export const validateRoutes = function (config) {
  return check.throw(check.objectOf(RouteConfigShape), config)
}

// eg:
// config = {
//   except: ['index']
// }
export const validateRoute = function (config) {
  return check.throw(RouteConfigShape, config)
}
