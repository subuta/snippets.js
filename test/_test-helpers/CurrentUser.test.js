/* global expect, describe, it */

import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'

import CurrentUser from 'lib/_test-helpers/CurrentUser'

import { Routes as RoutesConfig, Models as ModelsConfig } from 'test/fixtures/config'

import { generateSeeds } from 'lib/_utils/generateSeed'

describe('_test-helpers/CurrentUser', () => {
  it('should create currentUser payload', () => {
    expect(format(CurrentUser())).toMatchSnapshot()
  })

  it('should create currentUser payload with customSub', () => {
    const userSeed = _.first(generateSeeds(ModelsConfig.user, ModelsConfig.user.seeds))
    expect(format(CurrentUser(userSeed.auth0Id))).toMatchSnapshot()
  })
})
