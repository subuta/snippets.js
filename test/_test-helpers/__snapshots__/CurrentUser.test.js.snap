// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_test-helpers/CurrentUser should create currentUser payload 1`] = `
"export const createPayload = (sub) => ({
  sub,
  iss: 'https://xxx.com/',
  aud: ['https://xxx.com/api', 'https://xxx.auth0.com/userinfo'],
  scope: 'openid profile email'
})

export const currentUser = createPayload('google-oauth2|dummy')
"
`;

exports[`_test-helpers/CurrentUser should create currentUser payload with customSub 1`] = `
"export const createPayload = (sub) => ({
  sub,
  iss: 'https://xxx.com/',
  aud: ['https://xxx.com/api', 'https://xxx.auth0.com/userinfo'],
  scope: 'openid profile email'
})

export const currentUser = createPayload('turquoise revolutionize override')
"
`;
